### PR TITLE
Do not check whether tx is coinstake in PoW networks

### PR DESCRIPTION
**Pull request by yegorpetrov**
This bug prevents accepting btc testnet block 1439021, thereby freezing IBD at block 1439020. The [tx 55](https://live.blockcypher.com/btc-testnet/tx/ca6a2c79457300f669948d04bc1036330f8ded9d35d870c0854ac2c2de323218/) is the culprit.

### CODE REVIEW

**Comment by fassadlr** [comment]
@yegorpetrov So this is an issue when downloading BTC test net?

**Comment by yegorpetrov** [comment]
> @yegorpetrov So this is an issue when downloading BTC test net?

Yep, it's btc testnet. I didn't really try any mainnets, and the only other network I did try was litecoin testnet, which synced successfully prior to this fix (I haven't published LTC support code yet). And so with this fix, the btc testnet resumed syncing.

May be the definition of the `IsCoinStake` property is also a problem, but seeing how the transaction class does not have consensus context there, I don't see any better way to fix this.

**Comment by zeptin** [comment]
I can see why this occurs, although I think it may be better to separate the PoS-specific logic out of the rule into the PoS version instead of using `IsCoinStake` in a PoW context

**Comment by dangershony** [comment]
This is the condition for a POS trx
```
        public bool IsCoinStake
        {
            get
            {
                return this.Inputs.Any()
                    && !this.Inputs.First().PrevOut.IsNull
                    && this.Outputs.Count() >= 2
                    && this.Outputs.First().IsEmpty;
            }
        }
```
And for ouputs
```
        public bool IsEmpty => (this.Value == 0 && this.ScriptPubKey.Length == 0);
```

I always thought that an output with empty script and not value is not allowed on BTC

File: src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/CoinviewRule.cs [hunk 1]
```diff
@@ -79,7 +79,7 @@ public override async Task RunAsync(RuleContext context)
                     {
                         this.CheckInputs(tx, view, index.Height);
 
-                        if (!tx.IsCoinStake)
+                        if (!this.Consensus.IsProofOfStake || !tx.IsCoinStake)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/CoinviewRule.cs]: As @zeptin said I think my preference here is to create an abstract method maybe call it `GetTransactionFee` and pass in the `UnspentOutputSet` and the `Transaction` PoW implementation will just calculate the fee, POS will check the `IsCoinStake` condition and return zero if it is. 

`fees += this.GetTransactionFee(tx, view)`

Comment (yegorpetrov) [reply src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/CoinviewRule.cs]: Will do shortly.

File: src/Stratis.Bitcoin.Features.PoA/BasePoAFeatureConsensusRules/PoACoinviewRule.cs [hunk 2]
```diff
@@ -43,6 +44,11 @@ public override Money GetProofOfWorkReward(int height)
             return 0;
         }
 
+        protected override Money GetTransactionFee(UnspentOutputSet view, Transaction tx)
+        {
+            return view.GetValueIn(tx) - tx.TotalOut;
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.PoA/BasePoAFeatureConsensusRules/PoACoinviewRule.cs]: @noescape00 can you confirm?

Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA/BasePoAFeatureConsensusRules/PoACoinviewRule.cs]: Yes, looks correct

**Comment by codingupastorm** [comment]
Confirmed for smart contracts too. Will leave it to you to merge @dangershony.  Thanks @yegorpetrov

**Comment by dangershony** [comment]
I have synced testnet on both Stratis and Bitcoin this is good to go, however we need to merge conflicts.
@yegorpetrov if you manage to merge conflicts today please do otherwise I will to it later and push.

**Comment by yegorpetrov** [comment]
> I have synced testnet on both Stratis and Bitcoin this is good to go, however we need to merge conflicts.
> @yegorpetrov if you manage to merge conflicts today please do otherwise I will to it later and push.

@dangershony I've resolved the conflict.

**Pull request merged**